lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq/Gen.hs

```diff
@@ -0,0 +1,342 @@
+nextShrinkState txSeq = \case
+    ShrinkState phase (_ : actions) ->
+        Just $ ShrinkState phase actions
+    ShrinkState phase [] ->
+        Just $ case nextShrinkPhase phase of
+            Nothing -> ShrinkStateFinished
+            Just sp -> shrinkPhaseToState sp txSeq
+    ShrinkStateFinished ->
+        Nothing
+
+-- | Initializes a 'ShrinkState' for the given phase and partially shrunk
+--   sequence.
+--
+shrinkPhaseToState :: ShrinkPhase -> TxSeq -> ShrinkState
+shrinkPhaseToState phase txSeq =
+    ShrinkState phase (shrinkPhaseActions txSeq phase)
+
+getShrinkPhase :: ShrinkableTxSeq -> Maybe ShrinkPhase
+getShrinkPhase txSeq = case shrinkState txSeq of
+    ShrinkState phase _ -> Just phase
+    ShrinkStateFinished -> Nothing
+
+getShrinkState :: ShrinkableTxSeq -> ShrinkState
+getShrinkState = shrinkState
+
+-- | Generates a valid transaction for the given 'UTxO' set.
+--
+genTxFromUTxO :: Gen Address -> UTxO -> Gen Tx
+genTxFromUTxO genAddr u = do
+    (inputs, _) <-
+        selectUTxOEntries u =<< chooseInt (1, 2)
+    (collateralInputs, _) <-
+        selectUTxOEntries u =<< chooseInt (1, 2)
+    withdrawals <-
+        genMapWith genRewardAccount genCoinPositive
+    let inputValue = mconcat
+            [ F.foldMap (tokens . snd) inputs
+            , F.foldMap TokenBundle.fromCoin withdrawals
+            ]
+    let collateralInputValue =
+            F.foldMap (tokens . snd) collateralInputs
+    feeCoin <-
+        min (TokenBundle.coin inputValue) <$> chooseCoin (Coin 1, Coin 4)
+    let inputValueMinusFee =
+            inputValue `TokenBundle.difference` TokenBundle.fromCoin feeCoin
+    outputBundles <-
+        genTokenBundlePartitionNonNull inputValueMinusFee =<< chooseInt (1, 3)
+    collateralOutputBundles <-
+        elements [[], [collateralInputValue]]
+    outputAddresses <-
+        vectorOf (length outputBundles) genAddr
+    collateralOutputAddresses <-
+        vectorOf (length collateralOutputBundles) genAddr
+    scriptValidity <- elements
+        [ Nothing
+        , Just TxScriptValid
+        , Just TxScriptInvalid
+        ]
+    pure $ txWithoutIdToTx TxWithoutId
+        { fee =
+            Just feeCoin
+        , resolvedInputs =
+            fmap (TokenBundle.getCoin . tokens) <$> inputs
+        , resolvedCollateralInputs =
+            fmap (TokenBundle.getCoin . tokens) <$> collateralInputs
+        , outputs =
+            zipWith TxOut outputAddresses outputBundles
+        , collateralOutput = listToMaybe $
+            zipWith TxOut collateralOutputAddresses collateralOutputBundles
+        , metadata =
+            Nothing
+        , withdrawals
+        , scriptValidity
+        }
+
+--------------------------------------------------------------------------------
+-- Utilities
+--------------------------------------------------------------------------------
+
+boundedEnumSucc :: (Bounded a, Enum a, Ord a) => a -> Maybe a
+boundedEnumSucc a
+    | a >= maxBound = Nothing
```

Comment:
I don't understand how a  `Bounded a` could be greater than its max bound? 
```suggestion
    | a == maxBound = Nothing
```

Reply:
> I don't understand how a Bounded a could be greater than its max bound?

In the general case, it's quite possible for someone to define `Bounded` and `Ord` instances for some type `t`, where there exists some value `a :: t` such that `a > maxBound`. It wouldn't be good practice to do this, but the compiler certainly wouldn't prevent it.

This successor function is only well-defined for values satisfying `< maxBound`. Therefore, it's not well-defined for values satisfying `>= maxBound`.

I'm not sure I see the advantage for limiting the boundary check to `==` here. An alternative to returning `Nothing` would be to raise a run-time error. But I cannot see a clear advantage, unless you can see one?